### ISSUE TITLE
ZeroRedundancyOptimizer .join() context

### ISSUE THREAD

**Issue by amacati** [pytorch/pytorch#54318]
## 🚀 Feature
Implement a join context for the ZeroRedundancyOptimizer (and distributed optimizers in general).

## Motivation
Uneven inputs in DDP lead to a NCCL communication error due to nodes missing in the allreduce phase during gradient updates. This is resolved by running the training loop in an model.join() context, which shadows forward and backward passes by inserting communication operations where necessary. For the newly introduced ZeroRedundancyOptimizer there is no such context, therefore uneven inputs lead to NCCL crashes again. It would be great to have a similar shadowing functionality for the ZeroRedundancyOptimizer and all future distributed optimizers.

## Pitch

Add a .join() context similar to DistributedDataParallel.join() to distributed optimizers, specifically ZeroRedundancyOptimizer.

## Alternatives

I don't like the idea of opening two contexts for the training loop, but if the optimizer is supposed to be kept separate from the model, I don't see a way to put them into the same context.

## Additional context

I ran into this issue while training a DDP model on uneven inputs. Since training fails with the ZeroRedundancyOptimizer and DDP but not with regular optimizers, I strongly suspect what I described here is what's happening. There might be something I've missed, but given that my errors look exactly the same as running DDP without the join() context and the behaviour is to be expected, I'm fairly confident this is the issue.

**Comment by mrshenli** [comment]
cc @blefaudeux

**Comment by blefaudeux** [comment]
I think that I understand the need user wise, interesting. It's not completely obvious how to do that in practice though, but I suppose that the constraints are the same as DDP, so it should work. Adding a [link](https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html#DistributedDataParallel.join) for future reference

**Comment by blefaudeux** [comment]
@amacati would you have a reproduction case, to make sure that I understand what you mean with "uneven inputs" ? This would help implementing that in any case

**Comment by mrshenli** [comment]
Hey @blefaudeux, I think @amacati was referring to this [`DDP.join()`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.join) API. I agree that it can be an issue when say one trainer's dataloader produces 100 batches while another trainer's dataloader produces 101. As as result the first trainer won't match the collective communications used by ZeroRedundancyOptimizer in the 101th iteration on the second trainer. 

cc @rohan-varma does it make sense if we have general `join()` context that both DDP and the optimizer can interpret?

**Comment by rohan-varma** [comment]
Thanks for this feature request! I agree that the usefulness of `model.join()` would be limited in the case of using Zero redundancy as it would fail due to the collective comm. in the optimizer as you described. 

Currently the join() API only supports models/training loops without collective comm, other than the communication DDP does (allreduce, and some broadcasts etc to rebuild buckets). We are planning to address this limitation with: which will enable the join API to optionally throw a StopIteration when uneven inputs are detected, and then the user app can catch this error and decide how to proceed. This will work with models/training loops with collective comm such as zero. 

Although, to natively support this in the current mode might require additional design thought. For example, the reason `model.join()` works is that we are able to reliably shadow the communication that DDP does, because we know the tensor shapes that need to be passed into allreduce, for example. 

I haven't looked into zero too closely but the main thing to figure out is if we can reliably shadow the collective comm. the optimizer does. However, even if we can reliably shadow this, the shadowing would be "hardcoded" in the sense that it would not be easily extensible to other optimizers that shard in a different way, for example. Currently thinking through ideas on how to make this more generic, but open to suggestions.

**Comment by mrshenli** [comment]
Self-assigned to reverse it for a new team member
